store/tikv/gc_worker.go

```diff
@@ -0,0 +1,332 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tikv
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/tidb"
+)
+
+// GCWorker periodically triggers GC process on tikv server.
+type GCWorker struct {
+	uuid    string
+	store   *tikvStore
+	session tidb.Session
+	quit    chan struct{}
+	done    chan error
+}
+
+// NewGCWorker creates a GCWorker instance.
+func NewGCWorker(store *tikvStore) (*GCWorker, error) {
+	session, err := tidb.CreateSession(store)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	ver, err := store.CurrentVersion()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	worker := &GCWorker{
+		uuid:    fmt.Sprintf("gcworker_%d", ver.Ver),
+		store:   store,
+		session: session,
+		quit:    make(chan struct{}),
+		done:    make(chan error),
+	}
+	go worker.start(ver.Ver)
+	return worker, nil
+}
+
+// Close stops backgroud goroutines.
+func (w *GCWorker) Close() {
+	close(w.quit)
+}
+
+const (
+	gcWorkerTickInterval = time.Minute
+	gcWorkerLease        = time.Minute * 2
+	gcRunInterval        = time.Minute * 10
+	gcLeaderUUIDKey      = "tikv_gc_leader_uuid"
+	gcLeaderLeaseKey     = "tikv_gc_leader_lease"
+	gcLeaderLeaseFormat  = "20060102-15:04:05 -0700 MST"
+)
+
+func (w *GCWorker) start(ver uint64) {
+	log.Infof("[gc worker] %s start.", w.uuid)
+	safePoint := ver
+	gcIsRunning := false
+	ticker := time.NewTicker(gcWorkerTickInterval)
+	for {
+		select {
+		case <-ticker.C:
+			isLeader, err := w.checkLeader()
+			if err != nil {
+				log.Warnf("[gc worker] check leader err: %v", err)
+				break
+			}
+			if isLeader && !gcIsRunning {
+				if w.store.oracle.IsExpired(safePoint, uint64(gcRunInterval/time.Millisecond)) {
+					gcIsRunning = true
+					go w.runGCJob(safePoint)
+				}
+			}
+		case err := <-w.done:
+			gcIsRunning = false
+			if err != nil {
+				log.Errorf("[gc worker] runGCJob error: %v", err)
+				break
+			}
+			ver, err := w.store.CurrentVersion()
+			if err != nil {
+				log.Errorf("[gc worker] failed get current version: %v", err)
+				break
+			}
+			safePoint = ver.Ver
+		case <-w.quit:
+			log.Infof("[gc worker] (%s) quit.", w.uuid)
+			break
```

Comment:
If you want quit the start function, break can't do it. Break only breaks select here.